cmd/download.go

```diff
@@ -5,166 +5,258 @@
+	Kind          string
+	Version       string
+	Registry      bool
+	CloudProvider ProviderOptSpec
+	KindSpec      ProviderKind
+}
 
-	// Preparing all the necessary data for a worker pool
-	var wg sync.WaitGroup
-	var numberOfWorkers int
-	if parallel {
-		numberOfWorkers = runtime.NumCPU() + 1
+// newUrl initialize the PackageURL struct
+func newPackageURL(prefix string, blocks []string, kind, version string, registry bool, cloud ProviderOptSpec, kindSpec ProviderKind) *PackageURL {
+	return &PackageURL{
+		Prefix:        prefix,
+		Registry:      registry,
+		Blocks:        blocks,
+		Kind:          kind,
+		Version:       version,
+		CloudProvider: cloud,
+		KindSpec:      kindSpec,
+	}
+}
+
+// getConsumableURL returns an url that can be used for download
+func (n *PackageURL) getConsumableURL() string {
+
+	if !n.Registry {
+		return n.getURLFromCompanyRepos()
+	}
+
+	return fmt.Sprintf("%s/%s%s?ref=%s", n.KindSpec.pickCloudProviderURL(n.CloudProvider), n.Blocks[0], ".git", n.Version)
+
+}
+
+func (n *PackageURL) getURLFromCompanyRepos() string {
+	if len(n.Blocks) == 0 {
+		return ""
+	}
+
+	dG := ""
+
+	if strings.HasPrefix(n.Prefix, "git::https") {
+		dG = ".git"
+	}
+
+	if len(n.Blocks) == 1 {
+		return fmt.Sprintf("%s-%s%s//%s?ref=%s", n.Prefix, n.Blocks[0], dG, n.Kind, n.Version)
+	}
+
+	remainingBlocks := ""
+
+	for i := 1; i < len(n.Blocks); i++ {
+		remainingBlocks = path.Join(remainingBlocks, n.Blocks[i])
+	}
+
+	return fmt.Sprintf("%s-%s%s//%s/%s?ref=%s", n.Prefix, n.Blocks[0], dG, n.Kind, remainingBlocks, n.Version)
+
+}
+
+func downloadProcess(wg *sync.WaitGroup, opts DownloadOpts, data Package, errChan chan<- error, i int) {
+	// deferring the worker to be done
+	defer wg.Done()
+
+	logrus.Debugf("%d : received data %v", i, data)
+
+	// Checking git clone protocol
+	p := ""
+	if opts.Https {
+		p = httpsRepoPrefix
 	} else {
-		numberOfWorkers = 1
+		p = sshRepoPrefix
+	}
+
+	// Create the package URL from the data received to download the package
+	pU := newPackageURL(
+		p,
+		strings.Split(data.Name, "/"),
+		data.Kind,
+		data.Version,
+		data.Registry,
+		data.ProviderOpt,
+		data.ProviderKind)
+
+	u := pU.getConsumableURL()
+
+	downloadErr := get(u, data.Dir, getter.ClientModeDir, true)
+
+	// Checking if repository was found otherwise fallback to the old prefix, if fallback fails sends error to the error channel
+	if downloadErr != nil && strings.Contains(downloadErr.Error(), "Repository not found") {
+		o := humanReadableSource(pU.getConsumableURL())
+
+		if opts.Https {
+			pU.Prefix = fallbackHttpsRepoPrefix
+		} else {
+			pU.Prefix = fallbackSshRepoPrefix
+		}
+
+		logrus.Warningf("error downloading %s, falling back to %s", o, humanReadableSource(pU.getConsumableURL()))
+
+		downloadErr = get(pU.getConsumableURL(), data.Dir, getter.ClientModeDir, true)
+		if downloadErr != nil {
+			err := os.RemoveAll(data.Dir)
+			if err != nil {
+				logrus.Errorf("error removing directory %s: %s", data.Dir, err.Error())
+			}
```

Comment:
```suggestion
			if err := os.RemoveAll(data.Dir); err != nil {
				logrus.Errorf("error removing directory %s: %s", data.Dir, err.Error())
			}
```

Reply:
why is this error logged here and the downloadErr passed through an error channel and logger later?